src/main/java/com/pnu/ordermanagementapp/controller/ClientController.java

```diff
@@ -57,17 +70,51 @@ public String update(@PathVariable Long id, Model model, @AuthenticationPrincipa
     }
 
     @PostMapping("/new")
-    public String createClient(@ModelAttribute Client client, @AuthenticationPrincipal User user) {
+    public String createClient(@ModelAttribute("client") ClientFormSubmitDto clientFormDto, Model model,
+                               @AuthenticationPrincipal User user) {
+        if (validateClientFormDto(clientFormDto, model)) return "client/create_form";
+
+        Client client = Client.builder()
+                .firstName(clientFormDto.getFirstName())
+                .lastName(clientFormDto.getLastName())
+                .email(clientFormDto.getEmail())
+                .build();
+
         clientService.create(client, user.getId());
         return "redirect:/clients";
     }
 
     @PostMapping("/update")
-    public String updateClient(@ModelAttribute Client client, @AuthenticationPrincipal User user) {
+    public String updateClient(@ModelAttribute("client") ClientFormSubmitDto clientFormDto, Model model,
+                               @AuthenticationPrincipal User user) {
+
+        if (validateClientFormDto(clientFormDto, model)) return "client/update_form";
+
+        Client client = Client.builder()
+                .id(clientFormDto.getId())
+                .firstName(clientFormDto.getFirstName())
+                .lastName(clientFormDto.getLastName())
+                .email(clientFormDto.getEmail())
+                .build();
+
         clientService.update(client, user.getId());
         return "redirect:/clients";
     }
 
+    private boolean validateClientFormDto(@ModelAttribute("client") ClientFormSubmitDto clientFormDto, Model model) {
```

Comment:
Move this validations to separate component under `util` package.
Also, it should return only error message (and you can then check if it's empty) and do not have another side affects like mutation of `Model`. There are completely no point in setting `client` to `Model` here as well as `error`, both of them should be set in the controller, based on the result from validator component.

It would be even better if validator returned object like this:

```
@Data
@Builder
class ValidationResult {

    private String errorMessage;

    boolean isError() {
        return StringUtils.isNotBlank(errorMessage);
    }

    static ValidationResult noError() {
        return ValidationResult.builder().build();
    }

    static ValidationResult withError(String errorMessage) {
        return ValidationResult.builder()
                .errorMessage(errorMessage)
                .build();
    }

}
```